File: app/src/main/java/ru/strcss/projects/moneycalc/moneycalcandroid/componets/settings/SettingsPresenter.java
package ru.strcss.projects.moneycalc.moneycalcandroid.componets.settings;

import android.support.annotation.Nullable;

import javax.inject.Inject;
import javax.inject.Singleton;

import ru.strcss.projects.moneycalc.moneycalcandroid.api.MoneyCalcServerDAO;
import ru.strcss.projects.moneycalc.moneycalcandroid.storage.EventBus;
import ru.strcss.projects.moneycalc.moneycalcdto.dto.MoneyCalcRs;
import ru.strcss.projects.moneycalc.moneycalcdto.entities.SettingsLegacy;
import rx.Observer;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

import static ru.strcss.projects.moneycalc.moneycalcandroid.App.getAppContext;
import static ru.strcss.projects.moneycalc.moneycalcandroid.utils.ActivityUtils.snackBarAction;
import static ru.strcss.projects.moneycalc.moneycalcandroid.utils.events.Event.SETTING_UPDATED;
import static ru.strcss.projects.moneycalc.moneycalcandroid.utils.logic.ComponentsUtils.getErrorBodyMessage;

@Singleton
public class SettingsPresenter implements SettingsContract.Presenter {

    private final MoneyCalcServerDAO moneyCalcServerDAO;

    private final EventBus eventBus;

    @Inject
    SettingsPresenter(MoneyCalcServerDAO moneyCalcServerDAO, EventBus eventBus) {
        this.moneyCalcServerDAO = moneyCalcServerDAO;
        this.eventBus = eventBus;
    }

    @Nullable
    private SettingsContract.View view;

    @Override
    public void takeView(SettingsContract.View view) {
        this.view = view;
    }

    @Override
    public void dropView() {
        view = null;
    }

    @Override
    public void updateSettings(SettingsLegacy settings) {
        moneyCalcServerDAO.updateSettings(settings)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<MoneyCalcRs<SettingsLegacy>>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable ex) {
                        ex.printStackTrace();
                        snackBarAction(getAppContext(), getErrorBodyMessage(ex));
                        //                        String errorBodyMessage = getErrorBodyMessage((HttpException) ex);
//                        snackBarAction(view.getContext(), errorBodyMessage);
                    }

                    @Override
                    public void onNext(MoneyCalcRs<SettingsLegacy> updateRs) {
                        if (updateRs.isSuccessful()) {
                            eventBus.addEvent(SETTING_UPDATED);
                            view.showUpdateSuccess();
                        } else {
                            eventBus.addErrorEvent(updateRs.getMessage());
                            //                            view.showErrorMessage(updateRs.getMessage());
                        }
                    }
                });
    }
}

File: app/src/main/java/ru/strcss/projects/moneycalc/moneycalcandroid/componets/settings/SettingsPresenter.kt
package ru.strcss.projects.moneycalc.moneycalcandroid.componets.settings

import ru.strcss.projects.moneycalc.moneycalcandroid.App.getAppContext
import ru.strcss.projects.moneycalc.moneycalcandroid.api.MoneyCalcServerDAO
import ru.strcss.projects.moneycalc.moneycalcandroid.storage.EventBus
import ru.strcss.projects.moneycalc.moneycalcandroid.utils.ActivityUtils.snackBarAction
import ru.strcss.projects.moneycalc.moneycalcandroid.utils.events.Event.SETTING_UPDATED
import ru.strcss.projects.moneycalc.moneycalcandroid.utils.logic.ComponentsUtils.Companion.getErrorBodyMessage
import ru.strcss.projects.moneycalc.moneycalcdto.dto.MoneyCalcRs
import ru.strcss.projects.moneycalc.moneycalcdto.entities.SettingsLegacy
import rx.Observer
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SettingsPresenter @Inject
internal constructor(private val moneyCalcServerDAO: MoneyCalcServerDAO, private val eventBus: EventBus) : SettingsContract.Presenter {

    private var view: SettingsContract.View? = null

    override fun takeView(view: SettingsContract.View) {
        this.view = view
    }

    override fun dropView() {
        view = null
    }

    override fun updateSettings(settings: SettingsLegacy) {
        moneyCalcServerDAO.updateSettings(settings)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<MoneyCalcRs<SettingsLegacy>> {
                    override fun onCompleted() {}

                    override fun onError(ex: Throwable) {
                        ex.printStackTrace()
                        snackBarAction(getAppContext(), getErrorBodyMessage(ex))
                    }

                    override fun onNext(updateRs: MoneyCalcRs<SettingsLegacy>) {
                        if (updateRs.isSuccessful) {
                            eventBus.addEvent(SETTING_UPDATED)
                            view?.showUpdateSuccess()
                        } else {
                            eventBus.addErrorEvent(updateRs.message)
                        }
                    }
                })
    }
}
